File: src/test/java/com/fwdekker/randomness/array/ArraySettingsDialogTest.java
package com.fwdekker.randomness.array;

import com.intellij.openapi.ui.ValidationInfo;
import org.assertj.swing.edt.GuiActionRunner;
import org.assertj.swing.fixture.FrameFixture;
import org.assertj.swing.junit.testcase.AssertJSwingJUnitTestCase;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.swing.fixture.Containers.showInFrame;


/**
 * GUI tests for {@link ArraySettingsDialog}.
 */
public final class ArraySettingsDialogTest extends AssertJSwingJUnitTestCase {
    /**
     * An untouched {@code WordSettings} instance, thus having the default settings.
     */
    private static final ArraySettings DEFAULT_SETTINGS = new ArraySettings();

    private ArraySettings arraySettings;
    private ArraySettingsDialog arraySettingsDialog;
    private FrameFixture frame;


    @Override
    @SuppressWarnings("unchecked")
    protected void onSetUp() {
        arraySettings = new ArraySettings();
        arraySettingsDialog = GuiActionRunner.execute(() -> new ArraySettingsDialog(arraySettings));
        frame = showInFrame(robot(), arraySettingsDialog.createCenterPanel());
    }


    @Test
    public void testDefaultIsValid() {
        final ValidationInfo validationInfo = GuiActionRunner.execute(() -> arraySettingsDialog.doValidate());

        assertThat(validationInfo).isNull();
    }


    @Test
    public void testLoadSettingsCount() {
        frame.spinner("count").requireValue((long) DEFAULT_SETTINGS.getCount());
    }

    @Test
    public void testLoadSettingsBrackets() {
        frame.radioButton("bracketsSquare").requireSelected();
        frame.radioButton("bracketsCurly").requireNotSelected();
        frame.radioButton("bracketsRound").requireNotSelected();
    }

    @Test
    public void testLoadSettingsSeparator() {
        frame.radioButton("separatorComma").requireSelected();
        frame.radioButton("separatorSemicolon").requireNotSelected();
    }

    @Test
    public void testLoadSettingsSpaceAfterSeparator() {
        frame.checkBox("spaceAfterSeparator").requireSelected();
    }


    @Test
    public void testValidateCount() {
        GuiActionRunner.execute(() -> frame.spinner("count").target().setValue(983.24f));

        frame.spinner("count").requireValue(983L);
    }

    @Test
    public void testValidateCountNegative() {
        GuiActionRunner.execute(() -> frame.spinner("count").target().setValue(-172));

        final ValidationInfo validationInfo = arraySettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("count").target());
        assertThat(validationInfo.message).isEqualTo("Please enter a value greater than or equal to 1.");
    }

    @Test
    public void testValidateCountOverflow() {
        GuiActionRunner.execute(() -> frame.spinner("count").target().setValue((long) Integer.MAX_VALUE + 2L));

        final ValidationInfo validationInfo = arraySettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("count").target());
        assertThat(validationInfo.message).isEqualTo("Please enter a value less than or equal to 2147483647.");
    }


    @Test
    public void testSaveSettingsWithoutParse() {
        GuiActionRunner.execute(() -> {
            frame.spinner("count").target().setValue(642);
            frame.radioButton("bracketsCurly").target().setSelected(true);
            frame.radioButton("separatorSemicolon").target().setSelected(true);
            frame.checkBox("spaceAfterSeparator").target().setSelected(false);
        });

        arraySettingsDialog.saveSettings();

        assertThat(arraySettings.getCount()).isEqualTo(642);
        assertThat(arraySettings.getBrackets()).isEqualTo("{}");
        assertThat(arraySettings.getSeparator()).isEqualTo(";");
        assertThat(arraySettings.isSpaceAfterSeparator()).isEqualTo(false);
    }

    @Test
    public void testSaveSettingsWithParse() {
        frame.spinner("count").enterTextAndCommit("257");
        frame.radioButton("bracketsRound").check();
        frame.radioButton("separatorSemicolon").check();
        frame.checkBox("spaceAfterSeparator").uncheck();

        arraySettingsDialog.saveSettings();

        assertThat(arraySettings.getCount()).isEqualTo(257);
        assertThat(arraySettings.getBrackets()).isEqualTo("()");
        assertThat(arraySettings.getSeparator()).isEqualTo(";");
        assertThat(arraySettings.isSpaceAfterSeparator()).isEqualTo(false);
    }
}

File: src/test/kotlin/com/fwdekker/randomness/array/ArraySettingsDialogTest.kt
package com.fwdekker.randomness.array

import org.assertj.core.api.Assertions.assertThat
import org.assertj.swing.edt.FailOnThreadViolationRepaintManager
import org.assertj.swing.edt.GuiActionRunner
import org.assertj.swing.fixture.Containers.showInFrame
import org.assertj.swing.fixture.FrameFixture
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it


/**
 * GUI tests for [ArraySettingsDialog].
 */
object ArraySettingsDialogTest : Spek({
    lateinit var arraySettings: ArraySettings
    lateinit var arraySettingsDialog: ArraySettingsDialog
    lateinit var arraySettingsDialogConfigurable: ArraySettingsConfigurable
    lateinit var frame: FrameFixture


    beforeGroup {
        FailOnThreadViolationRepaintManager.install()
    }

    beforeEachTest {
        arraySettings = ArraySettings()
        arraySettings.count = 6
        arraySettings.brackets = "[]"
        arraySettings.separator = ","
        arraySettings.isSpaceAfterSeparator = false

        arraySettingsDialog = GuiActionRunner.execute<ArraySettingsDialog> { ArraySettingsDialog(arraySettings) }
        arraySettingsDialogConfigurable = ArraySettingsConfigurable(arraySettingsDialog)
        frame = showInFrame(arraySettingsDialog.getRootPane())
    }

    afterEachTest {
        frame.cleanUp()
    }


    describe("loading settings") {
        it("loads the settings' count") {
            frame.spinner("count").requireValue(6)
        }

        it("loads the settings' brackets") {
            frame.radioButton("bracketsNone").requireSelected(false)
            frame.radioButton("bracketsSquare").requireSelected(true)
            frame.radioButton("bracketsCurly").requireSelected(false)
            frame.radioButton("bracketsRound").requireSelected(false)
        }

        it("loads the settings' separator") {
            frame.radioButton("separatorComma").requireSelected(true)
            frame.radioButton("separatorSemicolon").requireSelected(false)
            frame.radioButton("separatorNewline").requireSelected(false)
        }

        it("loads the settings' settings for using a space after separator") {
            frame.checkBox("spaceAfterSeparator").requireSelected(false)
        }
    }

    describe("input handling") {
        it("truncates decimals in the count") {
            GuiActionRunner.execute { frame.spinner("count").target().value = 983.24f }

            frame.spinner("count").requireValue(983)
        }
    }

    describe("validation") {
        it("passes for the default settings") {
            GuiActionRunner.execute { arraySettingsDialog.loadSettings(ArraySettings()) }

            assertThat(arraySettingsDialog.doValidate()).isNull()
        }

        describe("count") {
            it("fails for a count of 0") {
                GuiActionRunner.execute { frame.spinner("count").target().value = 0 }

                val validationInfo = arraySettingsDialog.doValidate()

                assertThat(validationInfo).isNotNull()
                assertThat(validationInfo?.component).isEqualTo(frame.spinner("count").target())
                assertThat(validationInfo?.message).isEqualTo("Please enter a value greater than or equal to 1.")
            }

            it("passes for a count of 1") {
                GuiActionRunner.execute { frame.spinner("count").target().value = 1 }

                assertThat(arraySettingsDialog.doValidate()).isNull()
            }

            it("fails for a negative count") {
                GuiActionRunner.execute { frame.spinner("count").target().value = -172 }

                val validationInfo = arraySettingsDialog.doValidate()

                assertThat(validationInfo).isNotNull()
                assertThat(validationInfo?.component).isEqualTo(frame.spinner("count").target())
                assertThat(validationInfo?.message).isEqualTo("Please enter a value greater than or equal to 1.")
            }
        }
    }

    describe("saving settings") {
        it("correctly saves settings to a settings object") {
            GuiActionRunner.execute {
                frame.spinner("count").target().value = 642
                frame.radioButton("bracketsCurly").target().isSelected = true
                frame.radioButton("separatorSemicolon").target().isSelected = true
                frame.checkBox("spaceAfterSeparator").target().isSelected = false
            }

            arraySettingsDialog.saveSettings()

            assertThat(arraySettings.count).isEqualTo(642)
            assertThat(arraySettings.brackets).isEqualTo("{}")
            assertThat(arraySettings.separator).isEqualTo(";")
            assertThat(arraySettings.isSpaceAfterSeparator).isEqualTo(false)
        }
    }

    describe("configurable") {
        it("returns the correct display name") {
            assertThat(arraySettingsDialogConfigurable.displayName).isEqualTo("Arrays")
        }

        describe("modification detection") {
            it("is initially unmodified") {
                assertThat(arraySettingsDialogConfigurable.isModified).isFalse()
            }

            it("modifies a single detection") {
                GuiActionRunner.execute { frame.spinner("count").target().value = 124 }

                assertThat(arraySettingsDialogConfigurable.isModified).isTrue()
            }

            it("ignores an undone modification") {
                GuiActionRunner.execute { frame.spinner("count").target().value = 17 }
                GuiActionRunner.execute { frame.spinner("count").target().value = arraySettings.count }

                assertThat(arraySettingsDialogConfigurable.isModified).isFalse()
            }

            it("ignores saved modifications") {
                GuiActionRunner.execute { frame.spinner("count").target().value = 110 }

                arraySettingsDialogConfigurable.apply()

                assertThat(arraySettingsDialogConfigurable.isModified).isFalse()
            }
        }

        describe("resets") {
            it("resets all fields properly") {
                GuiActionRunner.execute {
                    frame.spinner("count").target().value = 642
                    frame.radioButton("bracketsCurly").target().isSelected = true
                    frame.radioButton("separatorSemicolon").target().isSelected = true
                    frame.checkBox("spaceAfterSeparator").target().isSelected = false

                    arraySettingsDialogConfigurable.reset()
                }

                assertThat(arraySettingsDialogConfigurable.isModified).isFalse()
            }
        }
    }
})
